internal/services/stock_exchange/algorithms.go

```diff
@@ -0,0 +1,340 @@
+		//Market  order
+		for _, elem := range buyBook {
+			if elem.OrderQuantity > sellOrderBody.Quantity {
+				ltp = elem.OrderPrice
+				config.DB.Exec("UPDATE buy_order_book SET order_quantity = ? , updated_at = ? WHERE id = ? ", elem.OrderQuantity-sellOrderBody.Quantity, time.Now(), elem.ID)
+				totalCost += elem.OrderPrice * sellOrderBody.Quantity
+				sellOrderBody.Quantity -= sellOrderBody.Quantity
+				UpdateLTP(ltp, sellOrderBody.StockName)
+				break
+			} else {
+				ltp = elem.OrderPrice
+				UpdateLTP(ltp, sellOrderBody.StockName)
+				totalCost += elem.OrderPrice * elem.OrderQuantity
+				sellOrderBody.Quantity -= elem.OrderQuantity
+				config.DB.Exec("DELETE FROM buy_order_book WHERE id = ?", elem.ID)
+				//config.DB.Delete(&model.SellOrderBook{}, elem.ID)
+			}
+		}
+
+		//update response
+		config.DB.Exec("DELETE FROM sell_order_book WHERE id = ?", id)
+		resp.Message = "Order Executed Successfully"
+		resp.Status = "Completed"
+		resp.OrderExecutionTime = time.Now()
+		resp.AveragePrice = totalCost / orderedQuantity
+		// order complete
+		orderResponse <- resp
+	}
+}
+
+// BuyOrder ...Update Buy Order actions on the StockExchange database
+func BuyOrder(buyOrderBody OrderRequest) (resp OrderResponse, err error) {
+
+	resp.Status = "Pending"
+	resp.OrderID = buyOrderBody.OrderID
+	resp.StockName = buyOrderBody.StockName
+	resp.Message = "Order Received, Pending"
+	newEntry := model.BuyOrderBook{
+		OrderID:           buyOrderBody.OrderID,
+		StockTickerSymbol: buyOrderBody.StockName,
+		OrderQuantity:     buyOrderBody.Quantity,
+		OrderStatus:       "Pending",
+		OrderPrice:        buyOrderBody.LimitPrice,
+		CreatedAt:         buyOrderBody.OrderPlacedTime,
+		UpdatedAt:         time.Now(),
+		//DeletedAt:         nil,
+	}
+	config.DB.Create(&newEntry)
+
+	newID := OrderId{}
+	err = config.DB.Raw("SELECT id FROM buy_order_book WHERE order_id = ?", buyOrderBody.OrderID).Scan(&newID).Error
+	if err != nil {
+		resp.Status = "Db fetch failed, order cancelled"
+		return resp, nil
+	}
+	go BuyOrderMatchingAlgo(buyOrderBody, resp, newID.id)
+	fmt.Println(<-orderResponse)
+	return resp, nil
+}
+
+// SellOrder ...Update Sell Order actions on the StockExchange database
+func SellOrder(sellOrderBody OrderRequest) (resp OrderResponse, err error) {
+
+	resp.Status = "Pending"
+	resp.OrderID = sellOrderBody.OrderID
+	resp.StockName = sellOrderBody.StockName
+	resp.Message = "Order Received, Pending"
+	newEntry := model.SellOrderBook{
+		OrderID:           sellOrderBody.OrderID,
+		StockTickerSymbol: sellOrderBody.StockName,
+		OrderQuantity:     sellOrderBody.Quantity,
+		OrderStatus:       "Pending",
+		OrderPrice:        sellOrderBody.LimitPrice,
+		CreatedAt:         sellOrderBody.OrderPlacedTime,
+		UpdatedAt:         time.Now(),
+		//DeletedAt:         nil,
+	}
+	config.DB.Create(&newEntry)
+	newID := OrderId{}
+	err = config.DB.Raw("SELECT id FROM sell_order_book WHERE order_id = ?", sellOrderBody.OrderID).Scan(&newID).Error
```

Comment:
Create a repo file and move this custom queries there. Also custom structs model can be placed there.

Reply:
why do we need to run this find query?

Reply:
changed